### PR TITLE
Added CSRF test to pass coverage on schema.py

### PR DESCRIPTION
**Pull request by sydoluciani**

### CODE REVIEW

**Comment by sydoluciani** [comment]
@stevepiercy coverage fails on this line of schema.py:
https://github.com/Pylons/deform/blob/master/deform/schema.py#L29

And the only way to pass the coverage is to call the fucntion with the same number of arguments, however invoking deferred_csrf_value directly, meaning invoking request object to get the CSRF token, which apparently creating a request object not easily possible in test environment and outside of the view. the request object needs to have a session and a CSRF token to pass the test, so created a mock version of request and session to return null instead of CSRF just to pass the test.

Now we get 100 percent coverage:
https://travis-ci.org/github/Pylons/deform/jobs/686742943#L306

Please check the code and let me know if it is acceptable, because this is the way going to implement test for renderer.py as well.

https://github.com/sydoluciani/deform/blob/drop-py27-py34/deform/tests/test_schema.py#L23

Thanks

File: deform/tests/test_schema.py [hunk 1]
```diff
@@ -20,6 +20,22 @@ def _makeOne(self):
...
+        kw = {"request": MagicMock(name="request")}
+        kw["request"].session = MagicMock(name="session")
+        kw["request"].session.get_csrf_token = MagicMock(
+            return_value=colander.null)
+
+        node = DummySchemaNode()
+        new_schema = CSRFSchema()
+        new_schema.default = deferred_csrf_value(node, kw)
+        self.assertEqual(new_schema.default, colander.null)
+
```
Comment (stevepiercy) [inline deform/tests/test_schema.py:38]: Would it be better to use an arbitrary value for the `return_value`? For example:

```
MOCK_CSRF_TOKEN = "SteveHasNoIdeaHowToFixThis"
...
            return_value=MOCK_CSRF_TOKEN)
...
            self.assertEqual(new_schema.default, MOCK_CSRF_TOKEN)
```

That would save an import of colander.

**Comment by stevepiercy** [comment]
@sydoluciani this looks good to me, but I would like to get a smarter person than myself to review it as well to provide guidance. I'm reaching out to a few colleagues, and we'll see. Thank you for your perseverance!

File: deform/tests/test_schema.py [hunk 2]
```diff
@@ -20,6 +20,20 @@ def _makeOne(self):
...
+    def test_deferred_csrf_value(self):
+        import colander
+        from deform.schema import CSRFSchema
+        from deform.schema import deferred_csrf_value
+        new_schema = CSRFSchema()
+        from unittest.mock import MagicMock
+        kw = {"request": colander.null}
+        kw["request"].session = MagicMock(name="session")
+        kw["request"].session.get_csrf_token = MagicMock(
+            return_value=colander.null)
```
Comment (sydoluciani) [inline deform/tests/test_schema.py:32]: That's true, we could return other value than colander.null, and remove the import colander, however collander.null is what being returned when function is decorated with @colander.deferred, until later on when the schema is bound to request, and that is the time kw['request'].session.get_csrf_token(node, kw) being accessed.
 
The way it is now is the closest to decorated function.

I wait for more feedbacks, but previous developers omitted schema.py and renderer.py for this reason that request object can't be instantiated easily outside of view, we could create a dummy version of request, but that would be the same as mocking it :)

Comment (stevepiercy) [reply deform/tests/test_schema.py:32]: Thanks for the explanation. I think your approach is good. I approved this PR.

I could merge this PR now or wait a day for additional comments. Either way, folks can always add comments. Please let me know. Thank you!

**Comment by stevepiercy** [comment]
I got a tip to try using [`pyramid.testing.DummyRequest`](https://github.com/Pylons/pyramid/blob/master/src/pyramid/testing.py#L273). [It's used throughout many Pylons Project projects](https://github.com/search?q=org%3APylons+DummyRequest&type=Code).

I might have time to try it this weekend. I got a couple of deadlines at the moment.

**Comment by sydoluciani** [comment]
That is good to know, it has session and get_csrf_token.
I should be able to complete testing renderer.py in weekend, then we can exchange notes.

**Comment by sydoluciani** [comment]
@stevepiercy 

Replaced MagicMock with DummyRequest:
https://github.com/sydoluciani/deform/blob/drop-py27-py34/deform/tests/test_schema.py#L28

Added test_renderer.py and used the example provided within the renderer.py.    
Currently testing if default_renderer isinstanceof ZPTRendererFactory.   
https://github.com/sydoluciani/deform/blob/drop-py27-py34/deform/tests/test_renderer.py#L23

Replaced bootstrap 3 files with bootstrap 4:   
bootstrap.min.css
bootstrap.min.js

**Comment by stevepiercy** [comment]
Have you signed https://github.com/Pylons/deform/blob/master/CONTRIBUTORS.txt yet?

**Comment by sydoluciani** [comment]
Just updated the CONTRIBUTORS.txt.

**Comment by stevepiercy** [comment]
@sydoluciani thank you for your work on this!

**Comment by sydoluciani** [comment]
> @sydoluciani thank you for your work on this!

Thank you for your guidance.
